src/host.cpp

```diff
@@ -216,43 +227,143 @@ namespace rhost {
+
+            if (!json[1].is<double>()) {
+                fatal_error("SetBlobSize: non-numeric blob Size");
+            }
+            auto size = static_cast<size_t>(json[1].get<double>());
+
             std::lock_guard<std::mutex> lock(blobs_mutex);
-            blobs.erase(blob_id);
+            auto& it = blobs.find(id);
+            if (it == blobs.end()) {
+                fatal_error("SetBlobSize: no blob with ID %llu", id);
+            }
+
+            it->second.resize(size);
+            respond_to_message(msg, static_cast<double>(it->second.size()));
         }
 
-        void destroy_blobs(const message& msg) {
-            assert(!strcmp(msg.name(), "!DestroyBlob"));
+        void read_blob(const message& msg) {
+            assert(!strcmp(msg.name(), "?ReadBlob"));
 
             auto json = msg.json();
+            if (!json[0].is<double>()) {
+                fatal_error("ReadBlob: non-numeric blob ID");
+            }
+            auto id = static_cast<blobs::blob_id>(json[0].get<double>());
+
+            if (!json[1].is<double>()) {
+                fatal_error("ReadBlob: non-numeric position");
+            }
+            long long pos = static_cast<long long>(json[1].get<double>());
+
+            if (!json[2].is<double>()) {
+                fatal_error("ReadBlob: non-numeric byte count");
+            }
+            long long count = static_cast<long long>(json[2].get<double>());
+
+            if (count < -1) {
+                fatal_error("ReadBlob: byte count cannot be < -1");
+            }
 
             std::lock_guard<std::mutex> lock(blobs_mutex);
-            for (auto val : json) {
-                if (!val.is<double>()) {
-                    fatal_error("DestroyBlob: non-numeric blob ID");
-                }
+            auto& it = blobs.find(id);
+            if (it == blobs.end()) {
+                fatal_error("ReadBlob: no blob with ID %llu", id);
+            }
 
-                auto id = static_cast<blobs::blob_id>(val.get<double>());
-                blobs.erase(id);
+            if (((pos != 0) && pos >= static_cast<long long>(it->second.size())) || pos < 0) {
+                blobs::blob empty;
+                respond_to_message(msg, empty);
+                return;
+            }
+
+            if (pos == 0 && count == -1) {
+                // Read all
+                respond_to_message(msg, it->second);
+                return;
+            } 
+            
+            // Read at position and count
+            size_t size = pos;
+            size += count;
+            if (count == -1 || size > it->second.size()) {
+                count = it->second.size() - pos;
+            }
+
+            blobs::blob::const_iterator begin = it->second.begin() + pos;
+            blobs::blob::const_iterator end = it->second.begin() + pos + count;
+
+            blobs::blob part(begin, end);
+            respond_to_message(msg, part);
+        }
+
+        void write_blob(const message& msg) {
```

Comment:
If I understand correctly, `ReadBlob` permits reads from arbitrary location, but `WriteBlob` doesn't? It feels like these should be consistent between each other (and if both allow offsets, you could expose it as a seekable stream).